base/strings/styled.jl

```diff
@@ -0,0 +1,363 @@
+"""
+    StyledString{S <: AbstractString} <: AbstractString
+
+A string annotated by properties (often styling information).
+
+More specifically, this is a thin wrapper around any other [`AbstractString`](@ref),
+which adds arbitrary named annotations to regions of the wrapped string.
+
+## Constructors
+
+```julia
+StyledString(s::S) -> StyledString{S}
+StyledString(s::S, props::Pair{Symbol, <:Any}...)
+StyledString(s::S, region::UnitRange{Int}, props::Pair{Symbol, <:Any}...)
+StyledString(s::S, properties::Vector{Tuple{UnitRange{Int}, Pair{Symbol, <:Any}}})
+```
+
+A StyledString can also be created with [`styledstring`](@ref), which acts much
+like [`string`](@ref) but preserves any styling present in the arguments.
+"""
+struct StyledString{S <: AbstractString} <: AbstractString
+    string::S
+    properties::Vector{Tuple{UnitRange{Int}, Pair{Symbol, Any}}}
```

Comment:
If we want to use this within Base, I think there cannot be an `Any` here so I think that needs to be restricted.

Reply:
Could you elaborate on why, is it a performance concern?

`Any` rather than say `Face` is a deliberate design choice to make this as broadly useful as possible. For instance, off the top of my head: it's useful to allow a `Face`, a `Symbol` naming a face, a `String` giving a hyperlink, Sixel image data, a source code position, and I'd imagine a bunch of things that will occur to other people.

Reply:
Bump. @KristofferC: can you elaborate on why this is an issue. Are you worried about invalidations at call sites?

Reply:
That was my worry yes (which I don't know if it is correct). But I was worried that we would use `StyledStrings` in a bunch of places and everywhere where `.properties` is used we would propagate an `Any` call to those methods and any method extension would then invalidate the whole printing system.

Maybe @timholy or @vtjnash knows better if this is a problem in practice.

Reply:
I think the key function in this regard may be `getface`, which is responsible for actually taking the string/some styles and saying "here's the overall/combined face you need to think about". It's taken some tinkering, but it's actually completely type stable.

```jl
julia> s = StyledString("some stuff", [(1:6, :face => :italic), (3:8, :face => :blue)])
"some stuff"

julia> @code_warntype Base.getface(s, 3)
MethodInstance for Base.getface(::StyledString{String}, ::Int64)
  from getface(s::StyledString, i::Integer) @ Base strings/faces.jl:424
Arguments
  #self#::Core.Const(Base.getface)
  s::StyledString{String}
  i::Int64
Body::Base.Face
1 ─ %1 = Base.textproperties(s, i)::Vector{Pair{Symbol, Any}}
│   %2 = Base.getface(%1)::Base.Face
└──      return %2
```

The other function called within that `getfaces` call and `eachstyled` is `textproperties`, which is also completely type stable.

```jl
julia> @code_warntype Base.textproperties(s, 3)
MethodInstance for Base.textproperties(::StyledString{String}, ::Int64)
  from textproperties(s::StyledString, i::Integer) @ Base strings/styled.jl:280
Arguments
  #self#::Core.Const(Base.textproperties)
  s::StyledString{String}
  i::Int64
Locals
  #485::Base.var"#485#486"{Int64}
  props::Vector{Tuple{UnitRange{Int64}, Pair{Symbol, Any}}}
Body::Vector{Pair{Symbol, Any}}
1 ─ %1 = Base.:(var"#485#486")::Core.Const(Base.var"#485#486")
│   %2 = Core.typeof(i)::Core.Const(Int64)
│   %3 = Core.apply_type(%1, %2)::Core.Const(Base.var"#485#486"{Int64})
│        (#485 = %new(%3, i))
│   %5 = #485::Base.var"#485#486"{Int64}
│   %6 = Base.getproperty(s, :properties)::Vector{Tuple{UnitRange{Int64}, Pair{Symbol, Any}}}
│        (props = Base.filter(%5, %6))
│   %8 = Base.broadcasted(Base.last, props)::Base.Broadcast.Broadcasted{Base.Broadcast.DefaultArrayStyle{1}, Nothing, typeof(last), Tuple{Vector{Tuple{UnitRange{Int64}, Pair{Symbol, Any}}}}}
│   %9 = Base.materialize(%8)::Vector{Pair{Symbol, Any}}
└──      return %9
```

This is the result of checking `@code_warntype` + benchmarking a bunch during development of these functions. I think `eachstyled` could maybe be improved a bit algorithmically, but microbenchmarks don't have me too concerned, though improvements (if possible) would be nice.

```jl
# This PR

julia> io = IOContext(IOBuffer(), :color => true)
IOContext(IOBuffer(data=UInt8[...], readable=true, writable=true, seekable=true, append=false, size=0, maxsize=Inf, ptr=1, mark=-1))

julia> @btime Base.banner(io)
  120.062 μs (1021 allocations: 180.61 KiB)

julia> io = IOContext(IOBuffer(), :color => false)
IOContext(IOBuffer(data=UInt8[...], readable=true, writable=true, seekable=true, append=false, size=0, maxsize=Inf, ptr=1, mark=-1))

julia> @btime Base.banner(io)
  12.746 μs (130 allocations: 17.77 KiB)

# On 1.10 beta 1

julia> @btime Base.banner(io)
  7.888 μs (31 allocations: 3.40 KiB)
```

So, that's a ~13x performance overhead for printing 60 styled regions. Only ~30% of this is actually from `eachstyle(banner_string)` though, which I take to mean that ~70% of this increace is in the printing, which is upstream of `strings/faces.jl` (see `strings/io.jl#979`).

Reply:
I'm going to mark this as resolved unless anybody wants to re-raise it :)